### PR TITLE
HIP activate callbacks

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- enforce HIP 4.0+
- update documentation
- update tests to use HIP-4.0.1 
- enable HIP unit test for callbacks
- remove the warning in CMake that HIP is experimental
- tests: use clang++ as host compiler

With the PR we require HIP 4.0, I dived into the commits of HIP and it looks as https://github.com/ROCm-Developer-Tools/HIP/issues/1205 is first fixed in HIP 4.0.

- [x] merge after #1270 (included as the first commit in this PR)

### CODE REVIEW

**Review by BenjaminW3 — DISMISSED** [review]

**Comment by psychocoderHPC** [comment]
I updated this PR:
- add install script to set environment and install dependency `rocrand/hipcrand` (not shipped with the compiler)
- use clang++ as host compiler (for HIP 4.0.1 it is clang 12)
- update test matrix in README.md

**Comment by SimeonEhrig** [comment]
> I updated this PR:
> 
>     * add install script to set environment and install dependency `rocrand/hipcrand` (not shipped with the compiler)
I think, this is a bug. `rocrand` is installed in the container: https://gitlab.com/hzdr/crp/alpaka-group-container/-/blob/master/generator/generator.py#L299

**Comment by psychocoderHPC** [comment]
> > I updated this PR:
> > ```
> > * add install script to set environment and install dependency `rocrand/hipcrand` (not shipped with the compiler)
> > ```
> 
> I think, this is a bug. `rocrand` is installed in the container: https://gitlab.com/hzdr/crp/alpaka-group-container/-/blob/master/generator/generator.py#L299

I checked and the **Ubuntu 20.04** container has no `rocrand` installed. Could be that it is installed in the Ubuntu 18.04 container but not in 20.04.

**Review by bernhardmgruber — COMMENTED** [review]
I cannot review/accept this due to lack of knowledge. Can someone else review it?

**Comment by SimeonEhrig** [comment]
> > > I updated this PR:
> > > ```
> > > * add install script to set environment and install dependency `rocrand/hipcrand` (not shipped with the compiler)
> > > ```
> > 
> > 
> > I think, this is a bug. `rocrand` is installed in the container: https://gitlab.com/hzdr/crp/alpaka-group-container/-/blob/master/generator/generator.py#L299
> 
> I checked and the **Ubuntu 20.04** container has no `rocrand` installed. Could be that it is installed in the Ubuntu 18.04 container but not in 20.04.

Sorry, I did a mistake. I didn't recognized, that this CI does not use our `alpaka-group-container`. In this case, you have to install `rocrand` manually.

**Comment by SimeonEhrig** [comment]
> I cannot review/accept this due to lack of knowledge. Can someone else review it?

I can review it. This PR is CI related, so it's fine for me.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -536,12 +536,12 @@ jobs:
...
+        - name: linux_hip_gcc-9_release_hip_only
+          os: ubuntu-20.04
+          env: {CXX: g++,     CC: gcc,    ALPAKA_CI_GCC_VER: 9,        ALPAKA_CI_STDLIB: libstdc++, CMAKE_BUILD_TYPE: Release, ALPAKA_CI_BOOST_BRANCH: boost-1.74.0, ALPAKA_CI_CMAKE_VER: 3.19.0,                     ALPAKA_CI_DOCKER_BASE_IMAGE_NAME: "rocm/dev-ubuntu-20.04:4.0.1", ALPAKA_ACC_GPU_HIP_ENABLE: ON, ALPAKA_ACC_GPU_HIP_ONLY_MODE: ON, ALPAKA_HIP_PLATFORM: clang, ALPAKA_CUDA_COMPILER: clang, ALPAKA_ACC_CPU_B_SEQ_T_FIBERS_ENABLE: OFF, ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLE: OFF, ALPAKA_ACC_CPU_B_SEQ_T_THREADS_ENABLE: OFF, ALPAKA_ACC_CPU_B_TBB_T_SEQ_ENABLE: OFF, ALPAKA_ACC_CPU_B_OMP2_T_SEQ_ENABLE: OFF, ALPAKA_ACC_CPU_B_SEQ_T_OMP2_ENABLE: OFF, ALPAKA_ACC_GPU_CUDA_ENABLE: OFF}
```
Comment (SimeonEhrig) [inline .github/workflows/ci.yml]: The name `gcc-9` and the `CXX: g++` are confusing. The build has nothing direct to do with the gcc version. You should change it to `clang++` and be sure, that it does not set the [wrong library path](https://github.com/alpaka-group/alpaka/blob/31544be047312c5554c99795386e327bf9f21ac9/script/run.sh#L40), if the HIP backend is enabled.

Don't forget this [line](https://github.com/alpaka-group/alpaka/blob/31544be047312c5554c99795386e327bf9f21ac9/script/run.sh#L98).

Comment (psychocoderHPC) [reply .github/workflows/ci.yml]: Yes I was also struggling with this line, but to install boost and Co we need a compiler `CXX: g++,     CC: gcc,    ALPAKA_CI_GCC_VER: 9,`. This means all dependencies will be installed with gcc and clang will be used for hipcc/clang++-12 will be used to build alpaka.
Is it OK for you to keep it?

Comment (SimeonEhrig) [reply .github/workflows/ci.yml]: Okay, then there is a new question. What happens in the jobs, where we use a vanila clang to compile CPU backends? Does the job also use g++ to compile boost and so one or use it clang? And if it use clang, why we can't use also the rocm clang to compile boost?

Comment (psychocoderHPC) [reply .github/workflows/ci.yml]: possible to use clang. I see you like to force me to switch to clang^^. I will do it.

Comment (SimeonEhrig) [reply .github/workflows/ci.yml]: Yes and no. I see just the confusion in the future. I did really bad experience in the last days ;-)

Comment (psychocoderHPC) [reply .github/workflows/ci.yml]: For unknown reason boost was not found when I use clang 10. I switched now to gcc 10 for the dependencies and hipcc(aka clang++-12) for alpaka. I added a comment to the yaml file which points to this mix.

Comment (SimeonEhrig) [reply .github/workflows/ci.yml]: I'm fine with this solution.

**Review by SimeonEhrig — APPROVED** [review]

**Comment by psychocoderHPC** [comment]
PR is now passing the CI, @alpaka-group/alpaka-maintainers please review this PR. THX @SimeonEhrig for the initial review.

**Review by bernhardmgruber — APPROVED** [review]
With somewhat limited understanding: looks good to me!

**Review by sbastrakov — APPROVED** [review]

**Pull request merged**